src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp

```diff
@@ -0,0 +1,1079 @@
+    {
+        return CHIP_IM_GLOBAL_STATUS(ConstraintError);
+    }
+
+    mPanMax = aPanMax;
+    MarkDirty(Attributes::PanMax::Id);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CameraAvSettingsUserLevelMgmtServer::SetZoomMax(uint8_t aZoomMax)
+{
+    if ((aZoomMax > kMaxZoomValue) || (aZoomMax < kMinZoomValue + 1))
+    {
+        return CHIP_IM_GLOBAL_STATUS(ConstraintError);
+    }
+
+    mZoomMax = aZoomMax;
+    MarkDirty(Attributes::ZoomMax::Id);
+    return CHIP_NO_ERROR;
+}
+
+// Mutators that may be invoked by a delegate in responding to command callbacks, or due to local on device changes
+//
+void CameraAvSettingsUserLevelMgmtServer::SetPan(Optional<int16_t> aPan)
+{
+    if (aPan.HasValue())
+    {
+        mMptzPosition.pan = aPan;
+        MarkDirty(Attributes::MPTZPosition::Id);
+    }
+}
+
+void CameraAvSettingsUserLevelMgmtServer::SetTilt(Optional<int16_t> aTilt)
+{
+    if (aTilt.HasValue())
+    {
+        mMptzPosition.tilt = aTilt;
+        MarkDirty(Attributes::MPTZPosition::Id);
+    }
+}
+
+void CameraAvSettingsUserLevelMgmtServer::SetZoom(Optional<uint8_t> aZoom)
+{
+    if (aZoom.HasValue())
+    {
+        mMptzPosition.zoom = aZoom;
+        MarkDirty(Attributes::MPTZPosition::Id);
+    }
+}
+
+void CameraAvSettingsUserLevelMgmtServer::AddMoveCapableVideoStreamID(uint16_t aVideoStreamID)
+{
+    // Make sure this isn't a duplicate ID, if not, add to the list
+    //
+    if (!KnownVideoStreamID(aVideoStreamID))
+    {
+        // Not found, this is a good add.
+        //
+        mDptzRelativeMove.push_back(aVideoStreamID);
+        MarkDirty(Attributes::DPTZRelativeMove::Id);
+    }
+}
+
+void CameraAvSettingsUserLevelMgmtServer::RemoveMoveCapableVideoStreamID(uint16_t aVideoStreamID)
+{
+    // Verify that this is a known ID, if it is, remove from the list
+    //
+    auto it = std::find(mDptzRelativeMove.begin(), mDptzRelativeMove.end(), aVideoStreamID);
+
+    if (it == mDptzRelativeMove.end())
+    {
+        ChipLogError(Zcl, "No matching video stream ID, removal not possible");
+        return;
+    }
+
+    mDptzRelativeMove.erase(it);
+    MarkDirty(Attributes::DPTZRelativeMove::Id);
+}
+
+bool CameraAvSettingsUserLevelMgmtServer::KnownVideoStreamID(uint16_t aVideoStreamID)
+{
+    // Verify that this is a known ID, if it is, remove from the list
+    //
+    auto it = std::find(mDptzRelativeMove.begin(), mDptzRelativeMove.end(), aVideoStreamID);
+
+    return (it == mDptzRelativeMove.end() ? false : true);
+}
+
+// Helper function for setting the next preset ID to use
+//
+void CameraAvSettingsUserLevelMgmtServer::UpdatePresetID()
+{
+
+    // Has the next possible incremented ID been used by a user set Preset?
+    //
+    uint8_t increment = 1;
+    do
+    {
+        auto it = std::find_if(mMptzPresetHelpers.begin(), mMptzPresetHelpers.end(), [=](const MPTZPresetHelper & mptzph) {
+            return mptzph.GetPresetID() == mCurrentPresetID + increment;
```

Comment:
So I am not sure I follow this logic.  Say mCurrentPresetID == 3, mMaxPresets == 5, the presets with 2, 3, 4, 5 are in use.  Can that situation happen?

In that case, this code will try check for a preset with id 4, then id 5, then id 6... will set mCurrentPresetID to 6 and return.  Is that the expected behavior?  Some documentation on mCurrentPresetID that explains the state management might be nice.

Reply:
Hmm, thinking about this the code isn't handling wrap around to zero when we hit an ID that is mMaxPresets -1.  This will need rework.

Reply:
Reworked the code and tested via chip-tool